### PR TITLE
Implementation of Top-KAST pruning algorithm

### PR DESCRIPTION
**Pull request by ohaijen**
Implementation of Top-KAST (Top-K always sparse training) pruning algorithm, published in https://arxiv.org/abs/2106.03517

The idea of the algorithm is that weights (forward step) and gradients (backward step) are pruned at two different sparsity levels (forward is sparser), allowing some weights to grow even as they are not used in the forward pass. Additionally, both forward-enabled and backward-enabled weights are decayed by the L2 norm at every step, allowing even weights that are not otherwise updated to potentially grow relative to those that are. The forward and backward masks are updated at each step in the original paper, but every ~100 steps is better for a batch size of 400.

Testing: added unit tests. In addition, ran a 100-epoch 90 forward/80 backward sparsity run, obtaining a 72.5% top-1 accuracy model. The paper reports 73.0% top-1 accuracy, but uses an impractically large batch size. Performance-wise, recomputing the mask at every step leads to a ~15% slowdown when training with a batch size of 400 on two Tesla V1000s,  which seems tolerable (especially because in practice we do it less often).

Command to run:
`CUDA_VISIBLE_DEVICES=2,3 WORLD_SIZE=2 python -m torch.distributed.launch --nproc_per_node=2 /home/jiofinova/sparseml/src/sparseml/pytorch/torchvision/train.py    --recipe="/home/jiofinova/resnet50-imagenet-topkast-recipe.yaml"     --dataset-path /nm/drive1/ILSVRC2012  --pretrained False     --arch-key resnet50     --batch-size 200    --workers 16  --opt sgd  --output-dir sparsification_example/resnet50-imagenet-topkast-90-50     --save-best-after 75`

topkast-recipe.yaml:
```
pruning_epochs: 100
finetuning_epochs: 0

# Learning rate hyperparams
init_lr: 0.25
final_lr: 0.


# Stabilization Stage
training_modifiers:
  - !EpochRangeModifier
    start_epoch: 0.0
    end_epoch: eval(stabilization_epochs + pruning_epochs + finetuning_epochs)

  - !SetLearningRateModifier
    start_epoch: 0.0
    learning_rate: eval(init_lr)
  - !LearningRateFunctionModifier
    init_lr: eval(init_lr)
    final_lr: eval(final_lr)
    lr_func: cosine
    start_epoch: 0
    end_epoch: 100

# Pruning Stage
pruning_modifiers:

  - !TopKASTPruningModifier
    forward_sparsity: 0.9
    backward_sparsity: 0.8
    update_frequency: 0.005
    start_epoch: eval(stabilization_epochs)
    end_epoch: eval(stabilization_epochs + pruning_epochs+finetuning_epochs)
    leave_enabled: True
    global_sparsity: False
    mask_type: unstructured
    params:
        - 're:.*sections.0.0.identity.conv.weight'
        - 're:.*sections.0.*.conv1.weight'
        - 're:.*sections.0.*.conv2.weight'
        - 're:.*sections.0.*.conv3.weight'
        - 're:.*sections.1.0.identity.conv.weight'
        - 're:.*sections.0.*.conv1.weight'
        - 're:.*sections.1.*.conv2.weight'
        - 're:.*sections.1.*.conv3.weight'
        - 're:.*sections.2.0.identity.conv.weight'
        - 're:.*sections.0.*.conv1.weight'
        - 're:.*sections.2.*.conv2.weight'
        - 're:.*sections.2.*.conv3.weight'
        - 're:.*sections.3.0.identity.conv.weight'
        - 're:.*sections.0.*.conv1.weight'
        - 're:.*sections.3.*.conv2.weight'
        - 're:.*sections.3.*.conv3.weight'

```

### CODE REVIEW

File: src/sparseml/pytorch/sparsification/pruning/modifier_pruning_topkast.py [hunk 1]
```diff
@@ -0,0 +1,424 @@
...
+        :return: mask creator object to be used by this pruning algorithm
+        """
+        return get_mask_creator_default(self.mask_type)
+
+    def check_mask_update(
+        self,
+        module: Module,
+        epoch: float,
+        steps_per_epoch: int,
+        recomputation_sparsity: Optional[float] = None,
```
Comment (natuan) [inline src/sparseml/pytorch/sparsification/pruning/modifier_pruning_topkast.py:234]: Add doc for recomputation_sparsity

**Review by natuan — APPROVED** [review]
LGTM (with some small changes)

File: src/sparseml/pytorch/sparsification/pruning/modifier_pruning_topkast.py [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+__all__ = ["TopKASTPruningModifier"]
+
+
+@PyTorchModifierYAML()
+class TopKASTPruningModifier(BasePruningModifier):
+    """
+    Implementation of
+    Top-KAST: Top-K Always Sparse Training:
+    https://arxiv.org/pdf/2106.03517.pdf.
+    AC/DC performs co-training of sparse and dense models, and can return both an
```
[the remark this answers is not in the document]
Comment (ohaijen) [reply src/sparseml/pytorch/sparsification/pruning/modifier_pruning_topkast.py:46]: Thanks, fixed.

**Review by natuan — APPROVED** [review]

**Review by anmarques — DISMISSED** [review]

**Review by anmarques — APPROVED** [review]

**Review by natuan — APPROVED** [review]
